homeassistant/components/wilight/switch.py

```diff
@@ -0,0 +1,460 @@
+"""Support for WiLight switches."""
+
+from pywilight.const import ITEM_SWITCH, SWITCH_PAUSE_VALVE, SWITCH_VALVE
+import voluptuous as vol
+
+from homeassistant.components.switch import ToggleEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv
+
+from . import DOMAIN, WiLightDevice
+from .support import wilight_to_hass_trigger, wilight_trigger as wl_trigger
+
+# Bitfield of features supported by the valve switch entities
+SUPPORT_WATERING_TIME = 1
+SUPPORT_PAUSE_TIME = 2
+SUPPORT_TRIGGER_1 = 4
+SUPPORT_TRIGGER_2 = 8
+SUPPORT_TRIGGER_3 = 16
+SUPPORT_TRIGGER_4 = 32
+
+# Attr of features supported by the valve switch entities
+ATTR_WATERING_TIME = "watering_time"
+ATTR_PAUSE_TIME = "pause_time"
+ATTR_TRIGGER_1 = "trigger_1"
+ATTR_TRIGGER_2 = "trigger_2"
+ATTR_TRIGGER_3 = "trigger_3"
+ATTR_TRIGGER_4 = "trigger_4"
+ATTR_TRIGGER_1_DESC = "trigger_1_description"
+ATTR_TRIGGER_2_DESC = "trigger_2_description"
+ATTR_TRIGGER_3_DESC = "trigger_3_description"
+ATTR_TRIGGER_4_DESC = "trigger_4_description"
+
+# Service of features supported by the valve switch entities
+SERVICE_SET_WATERING_TIME = "set_watering_time"
+SERVICE_SET_PAUSE_TIME = "set_pause_time"
+SERVICE_SET_TRIGGER_1 = "set_trigger_1"
+SERVICE_SET_TRIGGER_2 = "set_trigger_2"
+SERVICE_SET_TRIGGER_3 = "set_trigger_3"
+SERVICE_SET_TRIGGER_4 = "set_trigger_4"
+
+# Range of features supported by the valve switch entities
+RANGE_WATERING_TIME = 1800
+RANGE_PAUSE_TIME = 24
+
+# Service call validation schemas
+VALID_WATERING_TIME = vol.All(
+    vol.Coerce(int), vol.Range(min=1, max=RANGE_WATERING_TIME)
+)
+VALID_PAUSE_TIME = vol.All(vol.Coerce(int), vol.Range(min=1, max=RANGE_PAUSE_TIME))
+VALID_TRIGGER = wl_trigger
+
+# Service schemas
+SERVICE_SCHEMA_WATERING_TIME = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_WATERING_TIME): VALID_WATERING_TIME,
+    }
+)
+SERVICE_SCHEMA_PAUSE_TIME = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_PAUSE_TIME): VALID_PAUSE_TIME,
+    }
+)
+SERVICE_SCHEMA_TRIGGER_1 = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_TRIGGER_1): VALID_TRIGGER,
+    }
+)
+SERVICE_SCHEMA_TRIGGER_2 = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_TRIGGER_2): VALID_TRIGGER,
+    }
+)
+SERVICE_SCHEMA_TRIGGER_3 = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_TRIGGER_3): VALID_TRIGGER,
+    }
+)
+SERVICE_SCHEMA_TRIGGER_4 = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_TRIGGER_4): VALID_TRIGGER,
+    }
+)
```

Comment:
There is a lot of duplicate code here.  You could create these in a loop.

Reply:
This code really needs to be improved. I will make the modification.